libs/utils/energy_model.py

```diff
@@ -773,6 +773,216 @@ def _find_core_groups(cls, target):
+        #           cpulist-formatted representation of the cpus in the frequency domain
+        #        ../frequency
+        #           space-separated list of frequencies in corresponding order to capacities
+        #        ../power
+        #           space-separated list of power consumption in corresponding order to capacities
+        # taken together, the contents of capacity, frequency and power give you the required
+        # tuple for ActiveStates.
+        # hence, domain should be supplied as a glob, and fields should be
+        #   capacity, cpus, frequency, power
+
+        def sge_path(freq_domain, field):
+            f = '/sys/devices/system/cpu/energy_model/fd{}/{}'
+            return f.format(freq_domain, field)
+
+        # Read all the files we might need in one go, otherwise this will take
+        # ages.
+        sge_globs = [sge_path('**', 'capacity'),
+                     sge_path('**', 'cpus'),
+                     sge_path('**', 'frequency'),
+                     sge_path('**', 'power')]
+        sge_file_values = read_multiple_oneline_files(target, sge_globs)
+
+        if not sge_file_values:
+            raise TargetError('Energy Model not exposed in sysfs. '
+                              'Check CONFIG_SCHED_DEBUG is enabled.')
+        #for key in sge_file_values:
+        #    _log.info('file {}: {}'.format(key, sge_file_values[key]))
+
+        domain_match = re.compile('/sys/devices/system/cpu/energy_model/fd(\d+)/.*')
+        available_sge_freq_domains=[]
+        try:
+            domains = []
+            for key in sge_file_values:
+                m = domain_match.match(key)
+                if m:
+                    domain = m.group(1)
+                    if domain not in domains:
+                        domains.append(domain)
+            available_sge_freq_domains=domains
+        except:
+            pass
+        #_log.info("Available domains: {}".format(available_sge_freq_domains))
+
+        # convert a comma-separated list with ranges to a plain list
+        # e.g. '0,3-5' -> [ '0', '3', '4', '5' ]
+        def cpu_csl_to_list(cpu_csl):
+            result = []
+            for cpu_or_list in cpu_csl.split(','):
+                if '-' not in cpu_or_list:
+                    try:
+                        result.append(str(int(cpu_or_list)))
+                    except:
+                        raise RuntimeError('Unexpected value {} in cpu list string {}'.format(
+                                           cpu_or_list, cpu_csl))
+                else:
+                    list_cpus = cpu_or_list.split('-')
+                    if len(list_cpus) == 2:
+                        result.extend([str(c) for c in range(int(list_cpus[0]), int(list_cpus[1])+1)])
+            #_log.info('list {} converted to {}'.format(cpu_csl, result))
+            return result
+
+        # work out which frequency domain contains a particular CPU
+        def freq_domain_for_cpu(cpu):
```

Comment:
What about building a dict for that ? I don't know how to use that sge thing, so I'll use **read_tree_values** (it creates a dict) as an exemple:

``` python
sysfs_em = target.read_tree_values("/sys/devices/system/cpu/energy_model/", depth=3)

cpu_to_fdom = {}

for freq_domain, fields in sysfs_em.iteritems():
    cpus = ranges_to_list(fields["cpus"])
    for cpu in cpus:
        cpu_to_fdom[cpu] = freq_domain
    
```

Reply:
+1, using `read_tree_values()` sounds like the right method to parse the EM. We could actually use it to do more that just this CPU to fd mapping. We can probably convert the strings of "sysfs_em" into values and get a nice representation of the EM which can then be used directly to build the ActiveState objects. Based on your code, something as simple as:
```python
from devlib.utils.misc import *
from devlib.utils.types import *

sysfs_em = target.read_tree_values("/sys/devices/system/cpu/energy_model/", depth=3)
cpu_to_fdom = {}

for fd, fields in sysfs_em.iteritems():
    cpus = ranges_to_list(fields["cpus"])
    for cpu in cpus:
        cpu_to_fdom[cpu] = fd
    sysfs_em[fd]['cpus'] = cpus
    sysfs_em[fd]['frequency'] = map(integer, sysfs_em[fd]['frequency'].split(' '))
    sysfs_em[fd]['power'] =  map(integer, sysfs_em[fd]['power'].split(' '))
    sysfs_em[fd]['capacity'] = map(integer, sysfs_em[fd]['capacity'].split(' '))
``` 
gives the following `sysfs_em` struct (which has all we need, I think):

```
{'fd0': {'capacity': [133, 250, 351, 428, 462],
  'cpus': [0, 1, 2, 3],
  'frequency': [533000, 999000, 1402000, 1709000, 1844000],
  'power': [28, 70, 124, 187, 245]},
 'fd4': {'capacity': [391, 616, 782, 915, 1024],
  'cpus': [4, 5, 6, 7],
  'frequency': [903000, 1421000, 1805000, 2112000, 2362000],
  'power': [243, 500, 804, 1161, 1571]}}
```
And then,with this structure, `read_active_states()` can be simplified to something as simple as:
```python
def read_active_states(cpu, level):
    fd = sysfs_em[cpu_to_fdom[cpu]]
    cstates = zip(fd['capacity'], fd['power'])
    active_states = [ActiveState(c, p if not level else 0) for c, p in cstates]
    return OrderedDict(zip(sorted(fd['frequency']), active_states))
```